### PR TITLE
feat(SMTP): Add SSL support

### LINKED ISSUE #22492: TLS in SMTP is STARTTLS that is less secure than SSL(TLS)
Ref #4252

There is a wrong standpoint:

> Closing this as we are now on Django 1.11 which should support both SSL and TLS (and you really shouldn't be using SSL 🙂 )
_Originally posted by @BYK in https://github.com/getsentry/sentry/issues/4252#issuecomment-687329455_

We really should support SSL! Because the `SSL` of SMTP is include the well-known `TLS`. But in fact, the `TLS` of SMTP is `STARTTLS` instead of the well-known TLS.

According to [RFC3207](https://tools.ietf.org/html/rfc3207#section-11.1) (summary ref https://serverfault.com/a/648282), STARTTLS is less secure than TLS.

In the docs of most popular email service providers such as Tencent QQMail or Google Gmail, they both recommended using SSL (with port 465) rather than using STARTTLS (with port 587).

### PR DESCRIPTION
**Pull request by PMExtra**
Fixes #22492.

### CODE REVIEW

File: src/sentry/static/sentry/app/views/admin/options.tsx [hunk 1]
```diff
@@ -174,7 +174,13 @@ const definitions: Field[] = [
...
   {
     key: 'mail.use-tls',
-    label: t('Use TLS?'),
+    label: t('Use STARTTLS? (exclusive with SSL)'),
+    component: BooleanField,
+    defaultValue: () => false,
+  },
+  {
+    key: 'mail.use-ssl',
+    label: t('Use SSL? (exclusive with STARTTLS)'),
```
Comment (BYK) [inline src/sentry/static/sentry/app/views/admin/options.tsx:183]: I wonder if we can find a way to make these radio buttons: `Insecure`, `STARTTLS`, `SSL`.

@EvanPurkhiser @billyvg @priscilawebdev ideas?

**Comment by BYK** [comment]
@billyvg seems like we cannot run `Visual Snapshot` on PRs from forks again. Is this expected?

**Review by BYK — COMMENTED** [review]
This is one heck of a pull request, thanks a lot @PMExtra!

I'll wait for these builds and to hear back from our front-end folks before approving.

File: src/sentry/static/sentry/app/views/admin/adminMail.tsx [hunk 2]
```diff
@@ -80,11 +82,16 @@ export default class AdminMail extends AsyncView<{}, State> {
...
+          <dt>{t('STARTTLS?')}</dt>
           <dd>
             <pre className="val">{mailUseTls ? t('Yes') : t('No')}</pre>
           </dd>
 
+          <dt>{t('SSL?')}</dt>
+          <dd>
+            <pre className="val">{mailUseSsl ? t('Yes') : t('No')}</pre>
+          </dd>
+
```
Comment (PMExtra) [inline src/sentry/static/sentry/app/views/admin/adminMail.tsx:94]: There should be improved also.

File: src/sentry/static/sentry/app/views/admin/options.tsx (hunk 1 above)
Comment (billyvg) [reply src/sentry/static/sentry/app/views/admin/options.tsx:183]: It's doable, but I don't think there's a straightforward way to do it

**Comment by billyvg** [comment]
@BYK Sigh, not expected, GH docs are incomplete it seems

File: src/sentry/static/sentry/app/views/admin/options.tsx (hunk 1 above)
Comment (BYK) [reply src/sentry/static/sentry/app/views/admin/options.tsx:183]: @PMExtra - would you like to at least add some logic check for when both of these are checked or enabled?

**Comment by mattrobenolt** [comment]
Noting this here, since this introduces a new option in a hot path, this will need to be manually set in getsentry before we deploy this. Otherwise, we'll introduce a query every time an email is sent.

**Comment by billyvg** [comment]
@BYK Ok, snapshots are fixed, thanks for letting me know!

**Review by BYK — CHANGES_REQUESTED** [review]
@PMExtra sending this to your court for some basic logic checks for the accidental case where both `use-tls` and `use-ssl` are enabled.

Other than that the PR looks pretty good to me, thanks a ton!

**Comment by PMExtra** [comment]
@BYK Sorry, I'm not sure how can I do that. Django backend has this logic check and it will throw an exception if failed. If you means make a friendly notice in UI, sorry I'm not good at front-end.

**Review by BYK — APPROVED** [review]
>@BYK Sorry, I'm not sure how can I do that. Django backend has this logic check and it will throw an exception if failed. If you means make a friendly notice in UI, sorry I'm not good at front-end.

Ah, didn't know Django did this already. No UI needed, thanks a lot!

**Comment by PMExtra** [comment]
@BYK You're welcome. It's my pleasure to make Sentry greater.
It's a reference about Django check the SMTP options: https://github.com/django/django/blob/master/django/core/mail/backends/smtp.py#L30-L33

**Pull request merged**